pkg/lib/sms/sendcloud/smsclient.go

```diff
@@ -138,12 +139,42 @@ func (n *SendCloudClient) Send(ctx context.Context, options *smsclient.SendOptio
 		return &smsclient.SendResultSuccess{
 			DumpedResponse: dumpedResponse,
 		}, nil
+	} else {
+		return nil, n.makeError(sendResponse.StatusCode, dumpedResponse)
 	}
+}
 
-	// Failed case.
-	return nil, &smsclient.SendResultError{
+func (t *SendCloudClient) makeError(
+	statusCode int,
+	dumpedResponse []byte,
+) *smsclient.SendResultError {
+	err := &smsclient.SendResultError{
 		DumpedResponse: dumpedResponse,
 	}
+
+	// See https://www.sendcloud.net/doc/sms/api/
+	switch statusCode {
+	case 412:
+		err.Code = api.CodeInvalidPhoneNumber
+		err.ErrorDetail = fmt.Sprintf("%d", statusCode)
+	case 50000:
+		err.Code = api.CodeRateLimited
+		err.ErrorDetail = fmt.Sprintf("%d", statusCode)
+	case 422:
+		fallthrough
+	case 471:
+		fallthrough
+	case 474:
+		err.Code = api.CodeAuthenticationFailed
+		err.ErrorDetail = fmt.Sprintf("%d", statusCode)
+	case 499:
+		fallthrough
+	case 473:
+		err.Code = api.CodeAuthorizationFailed
```

Comment:
We should introduce api.CodeInsufficientBalance, and this should replace CodeAuthorizationFailed.

Reply:
CodeAuthorizationFailed may not be due to insufficient balance, but could be due to account suspended (Which can be due to insufficent balance).
So I think it is difficult to distinguish the real cause, let's just return authorization failed which means: The user exist, but it cannot be used to send the sms.
As we have ErrorDetail, the developer can find out the reason easily.

Reply:
Then CodeSuspendedAccount is preferred. CodeAuthorizationFailed is too vague.

Reply:
As discussed it will be change to DelieveryRejected